jerry-core/ecma/operations/ecma-regexp-object.c

```diff
@@ -1606,6 +1606,527 @@ ecma_regexp_search_helper (ecma_value_t regexp_arg, /**< regexp argument */
+    const lit_utf8_byte_t *const matched_p = ecma_regexp_match (&re_ctx, bc_start_p, current_str_p);
+
+    if (ECMA_RE_STACK_LIMIT_REACHED (matched_p))
+    {
+      result = ecma_raise_range_error (ECMA_ERR_MSG ("Stack limit exceeded."));
+      goto cleanup_array;
+    }
+
+    if (matched_p == NULL || matched_p == previous_str_p)
+    {
+      lit_utf8_incr (&current_str_p);
+      continue;
+    }
+
+    /* 13.c.iii.1. */
+    ecma_string_t *const str_p = ecma_new_ecma_string_from_utf8 (previous_str_p,
+                                                                 (lit_utf8_size_t) (current_str_p - previous_str_p));
+
+    result = ecma_builtin_helper_def_prop_by_index (array_p,
+                                                    array_length++,
+                                                    ecma_make_string_value (str_p),
+                                                    ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
+    JERRY_ASSERT (ecma_is_value_true (result));
+    ecma_deref_ecma_string (str_p);
+
+    if (array_length == limit)
+    {
+      result = array;
+      goto cleanup_context;
+    }
+
+    /* 13.c.iii.5. */
+    previous_str_p = matched_p;
+
+    uint32_t index = 1;
+    while (index < re_ctx.captures_count)
+    {
+      const ecma_value_t capture = ecma_regexp_get_capture_value (re_ctx.captures_p + index);
+      result = ecma_builtin_helper_def_prop_by_index (array_p,
+                                                      array_length++,
+                                                      capture,
+                                                      ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
+      JERRY_ASSERT (ecma_is_value_true (result));
+      ecma_free_value (capture);
+
+      if (array_length == limit)
+      {
+        result = array;
+        goto cleanup_context;
+      }
+
+      index++;
+    }
+
+    /* 13.c.iii.8. */
+    current_str_p = matched_p;
+  }
+
+  ecma_string_t *const str_p = ecma_new_ecma_string_from_utf8 (previous_str_p,
+                                                               (lit_utf8_size_t) (string_end_p - previous_str_p));
+
+  result = ecma_builtin_helper_def_prop_by_index (array_p,
+                                                  array_length++,
+                                                  ecma_make_string_value (str_p),
+                                                  ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
+  JERRY_ASSERT (ecma_is_value_true (result));
+  ecma_deref_ecma_string (str_p);
+
+  result = array;
+  goto cleanup_context;
+
+cleanup_array:
+  ecma_deref_object (array_p);
+cleanup_context:
+  ecma_regexp_cleanup_context (&re_ctx);
+cleanup_buffer:
+  if (string_flags & ECMA_STRING_FLAG_MUST_BE_FREED)
+  {
+    jmem_heap_free_block ((void *) string_buffer_p, string_size);
+  }
+cleanup_string:
+  ecma_deref_ecma_string (string_p);
+
+  return result;
+#endif /* ENABLED (JERRY_ES2015) */
```

Comment:
Looks hard to maintain, isn't there a better way than maintaining two large code paths?

Reply:
Unfortunately, ES6 uses an entirely different approach to RegExp based splitting, and the two are incompatible with eachother. ES6 requires us to reconstruct the RegExp with a sticky flag and split the string by adjusting the lastIndex property, while ES5 calls the builtin RegExp matching directly.

Maybe 10% of the code could be shared between the two implementations, but that would require a lot of guards to pull off, and in my opinion would just be messy.